Heres a new pull request thingy if you want

litehed opened the pull request.

litehed committed:
Added thingy to make testing field centric easier

couldnt test shooter code today sorry bb

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOpMain.java:
```diff
@@ -48,7 +49,6 @@
     private Button shooterStart, shooterStop, dpadUp, dpadDown, goalLift, toggleIntake;
-
     @Override
     public void initialize() {
```
Change the ConditionalCommand binding to `whileHeld()` instead of `whenPressed()`

litehed committed:
Changed to whileHeld in the class to test shooter

Ill change it on everything once I test

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jacksonSama/ShooterOnly.java:
```diff
@@ -36,13 +37,14 @@
 
     public GamepadEx m_driverOp, m_toolOp;
```
```suggestion
    private GamepadEx m_driverOp, m_toolOp;
```

JIceberg committed:
Change visibility field of gamepad objects to private

litehed committed:
peepeepoopoo check

peeeeeeeee

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jacksonSama/ShooterOnly.java:
```diff
@@ -73,13 +75,13 @@ public void initialize() {
         stopCommand = new Com_NoShoot(shooterSystem);
         toggleShooter = new GamepadButton(m_driverOp, GamepadKeys.Button.A)
-                .whenPressed(new ConditionalCommand(shootCommand, stopCommand, shooterSystem::active))
+                .whileHeld(new ConditionalCommand(shootCommand, stopCommand, shooterSystem::active))
                 .whenReleased(new InstantCommand(shooterSystem::toggle));
```
Idea: we use the `toggleWhenPressed` binding instead.
```suggestion
                .toggleWhenPressed(shootCommand);
```
Then we can remove the active and toggle methods from the Subsystem and also not longer need the stop command.

JIceberg committed:
Change conditional binding to built-in toggling in the Trigger

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/jacksonSama/ShooterOnly.java:
```diff
@@ -73,14 +75,13 @@ public void initialize() {
         toggleShooter = new GamepadButton(m_driverOp, GamepadKeys.Button.A)
-                .whenPressed(new ConditionalCommand(shootCommand, stopCommand, shooterSystem::active))
-                .whenReleased(new InstantCommand(shooterSystem::toggle));
+                .toggleWhenPressed(shootCommand);
```
In the `Com_Shoot` class, override the `cancel()` method:
```java
@Override
void cancel() {
    shooterSubsystem.stop(); /* or whatever the name of that private instance is */
    CommandScheduler.getInstance().cancel(this);
}
```

litehed committed:
takes poop on Jackson

litehed committed:
Got rid of those gross MotorExs and disabled stuff

litehed committed:
Vision Command Testing

welp I hope this works but who knows knowing me it probably doesnt

litehed committed:
fixes

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/VisionSystem.java:
```diff
@@ -0,0 +1,41 @@
+
+    public Size getStackSize(){
+        UGRectDetector.Stack stack = ugRectDetector.getStack();
+        switch (stack) {
+            case ZERO:
+                telemetry.addData("Rings Detected", "Zero");
+                telemetry.update();
+                return Size.ZERO;
+            case ONE:
+                telemetry.addData("Rings Detected", "One");
+                telemetry.update();
+                return Size.ONE;
+            case FOUR:
+                telemetry.addData("Rings Detected", "Four");
+                telemetry.update();
+                return Size.FOUR;
+            default:
+                break;
```
NICE! You got the hang of it. This is perfect.

litehed replied:
**YES WOOOOOOOOOO**

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutonomousKanye.java:
```diff
@@ -0,0 +1,62 @@
+        SequentialCommandGroup wobbleGoal = new SequentialCommandGroup(
+                visionCommand,
+                new SelectCommand(new HashMap<Object, Command>() {{
+                    put(VisionSystem.Size.ZERO, new InstantCommand(() -> {test.set(1);}));
+                    put(VisionSystem.Size.ONE, new InstantCommand(() -> {test.set(0.5);}));
+                    put(VisionSystem.Size.FOUR, new InstantCommand(() -> {test.set(0.1);}));
```
You can remove the curly brackets and the semicolons in the instant command lambda.

litehed replied:
dammit so close

litehed committed:
I had added the semicolons originally then when I didnt work I was like ooh brackets

JIceberg commented:
Okay that's all I'm going to help with for now. I want you to learn how to do the rest on your own. I'll just help if something breaks. Let's see how much you can get done/working.

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/Com_Vision.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.firstinspires.ftc.teamcode.subsystems.VisionSystem;
+
+public class Com_Vision extends CommandBase {
```
Whoops! You need to add an isFinished here for the sequential command group. Just override it and automatically return true. That way it only runs once. You should actually make this like an instant command and have the initialize method read the current stack height instead of the execute.

JIceberg commented:
Also make sure that for your drive commands when you drive to the wobble goal position and drive back, you override the `end()` method.

JIceberg commented:
The reason you want to override `end()` is because that way you can define the end behavior before the sequential command group moves to the next command in the sequence. `isFinished()` determines whether a command in the group should be ended and the next command should be run.

JIceberg commented:
Here's an example for how you would override that `end(boolean interrupted)` method for the drivebase.
```java
/* You should pass in an elapsed time object to the constructor */

@Override
public void initialize() {
    timer.reset();
}

@Override
public void execute() {
    // drive forward - use the drive subsystem (pass it into the constructor)
}

@Override
public void end(boolean interrupted) {
    // stop the drive - write a stop method for the drive subsystem and use that
}

@Override
public boolean isFinished() {
    return timer.milliseconds() < m_desiredTime;
}
```
Something like that. Just a template, don't copy it exactly.

litehed committed:
Basic Drive For Time Command

I think this is right I have faith

litehed committed:
Also an example of the DriveTimes use

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/Com_DriveTime.java:
```diff
@@ -0,0 +1,43 @@
+    private final ElapsedTime timey;
+    private final Double timeLength;
+
+    public Com_DriveTime(DriveSystem subsystem, Double strafeSpeed,
+                         Double forwardSpeed, Double turnSpeed, ElapsedTime time, Double amnt){
```
Why are these Double wrapper objects? Just make them doubles or DoubleSuppliers.

litehed committed:
fixed that double thing lol

peepee

litehed committed:
I made something for rishab which probably doesnt work

litehed committed:
Rotation Command Added

I created a command whihc will just make the robot rotate a selected number of degrees in any direction

litehed committed:
Notice me senpai help is required

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/drive/Com_Rotate.java:
```diff
@@ -0,0 +1,38 @@
+    //thx -past me
+    public Com_Rotate(DriveSystem subby, RevIMU revIMU, int degreesIn){
+        driveSystem = subby;
+        imu = revIMU;
+        //Question to Jackson I did this because I dont know what the imus max value is so I just
+        //said if it exceeds 360 subtract 360, does it have a max value is this needed I dont know
+        degrees = (degreesIn + imu.getHeading()) >= 360 ? (degreesIn + imu.getHeading())-360 : degreesIn + imu.getHeading();
```
It should just be (-180, 180). If you want you can call `imu.getRotation2d().getDegrees()` which will wrap it for you.

litehed replied:
ok thx I was having trouble with this

litehed replied:
wait but degrees still has the potential to exceed 180 so I dont see how the getRotation2d would help

JIceberg replied:
Creating the Rotation2d object clamps the angle in the constructor to -180 to 180.

litehed replied:
Ahhh that's interesting thanks

litehed committed:
poopoopeepoo

litehed committed:
BIG COMMIT

CHUNGA CHUNGA

litehed committed:
poopoopeepoopoopee

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutonomousKanye.java:
```diff
@@ -0,0 +1,91 @@
+        mecDrive = new DriveSystem(fL, fR, bL, bR);
+        visionSystem = new VisionSystem(ugRectDetector, telemetry);
+        visionCommand = new Com_Vision(visionSystem);
+        register(mecDrive, new SubsystemBase(){
+            @Override
+            public void periodic() {
+                telemetry.addData("imu heading", imu.getHeading());
+                telemetry.update();
+            }
+        });
```
Why?

litehed committed:
Reasons Why I hate Jackson

hi

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutonomousKanye.java:
```diff
@@ -0,0 +1,113 @@
+                intialStrafe,
+                visionCommand,
+                new SelectCommand(new HashMap<Object, Command>() {{
+                    put(VisionSystem.Size.ZERO, new GroupZero(mecDrive, time, voltageSensor, imu));
+                    put(VisionSystem.Size.ONE, new GroupOne(mecDrive, time, voltageSensor));
+                    put(VisionSystem.Size.FOUR, new GroupFour(mecDrive, time, voltageSensor, imu));
```
```java
put(VisionSystem.Size.ZERO, new ScheduleCommand(new GroupZero(mecDrive, time, voltageSensor, imu)));
```

litehed committed:
u suk

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/groups/GroupFour.java:
```diff
@@ -0,0 +1,25 @@
+    public GroupFour(DriveSystem drive, ElapsedTime time, VoltageSensor voltageSensor, RevIMU imu, WobbleSystem wobbleSystem) {
+        addCommands(
+                new Com_DriveTime(drive,0.5, 0D, 0D, time, 2.0),
+                new Com_DriveTime(drive, (12/voltageSensor.getVoltage())*-0.55, 0D, 0D, time, 3.7),
```
```suggestion
                new Com_DriveTime(drive, (12/voltageSensor.getVoltage())*-0.55, 0D, 0D, time, 3.4),
```

JIceberg commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/commands/groups/GroupFour.java:
```diff
@@ -0,0 +1,25 @@
+                new Com_DriveTime(drive, 0D, (12/voltageSensor.getVoltage())*-0.55, 0D, time, 7.5),
+                new Com_Rotate(drive, imu, 180),
+                new Com_PutDown(wobbleSystem, time),
+                new Com_DriveTime(drive, 0D, (12/voltageSensor.getVoltage())*-0.55, 0D, time, 4)
```
```suggestion
                new Com_DriveTime(drive, 0D, (12/voltageSensor.getVoltage())*-0.55, 0D, time, 3)
```

JIceberg committed:
change time to smaller amount for strafing right

JIceberg committed:
decrease backup time

JIceberg commented:
@liteheded Push your current code so we can merge into master (since this is code we competed with). Don't delete the branch this time so we can save progress.

litehed committed:
Code has been pushed

I just had a super cool idea that ima try out later for wobble grabber
